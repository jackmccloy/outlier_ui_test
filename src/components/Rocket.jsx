import React from 'react';

import PropTypes from 'prop-types';

import { Link } from 'react-router-dom';

const Rocket = ({ rocket }) => (
  <li>
    <h2>
      <Link to={`/rockets/${rocket.rocket_id}`}>
        { rocket.rocket_name }
      </Link>
    </h2>
  </li>
);

Rocket.propTypes = {
  rocket: PropTypes.shape({
    // Todo: This is not the full description of the data - it only includes
    // the fields that we currently use.
    rocket_name: PropTypes.string,
  }).isRequired,
}

export default Rocket;