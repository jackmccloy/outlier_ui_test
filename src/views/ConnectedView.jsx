import React, { Component } from 'react';
import { connect } from "react-redux";

import Navigation from '../components/Navigation';

const menu = Navigation();

function MasterLayoutHOC(WrappedComponent, pageName) {
  class MasterLayoutImpl extends Component {
    render() {

      const layoutProps = {
        menu,
        pageName
      };

      return (
        <WrappedComponent {...this.props} />
      );
    }
  }

  const mapStateToProps = state => state;

  const mapDispatchToProps = dispatch => ({
    dispatch
  });

  return connect(
    mapStateToProps,
    mapDispatchToProps
  )(MasterLayoutImpl);
}

export default MasterLayoutHOC;
