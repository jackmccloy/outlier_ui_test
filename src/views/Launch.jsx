import React, { Component } from 'react';

import { Link } from 'react-router-dom';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getLaunchesThunk } from "../actions/launches";
import ImagesArea from '../components/ImagesArea';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const LaunchH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class LaunchView extends Component {
  componentDidMount() {
    const { getLaunches } = this.props;
    getLaunches();
  }

  render() {
    const {
      match,
      launchCollection,
    } = this.props;

    if (!launchCollection || launchCollection.fetching) {
      return (
        <Section>
          <LaunchH1> SpaceX </LaunchH1>
          <div> LOADING... </div>
        </Section>
      );
    }

    if (launchCollection.error) {
      return (
        <Section>
          <LaunchH1> SpaceX </LaunchH1>
          <div> { launchCollection.error } </div>
        </Section>
      );
    }

    const launchId = parseInt(match.params.flight_number, 10);
    const thisLaunch = launchCollection.launches.filter(launch => launch.flight_number === launchId);

    if (thisLaunch.length !== 1) {
      return (
        <Section>
          <LaunchH1> SpaceX </LaunchH1>
          <div> NO DATA </div>
        </Section>
      );
    }

    return (
      <Section>
        <LaunchH1> SpaceX { thisLaunch[0].mission_name }, {thisLaunch[0].launch_year}</LaunchH1>
        <h2> Rocket: <Link to={`/rockets/${thisLaunch[0].rocket.rocket_id}`}>
          { thisLaunch[0].rocket.rocket_name }</Link>
        </h2>
        <div> { thisLaunch[0].details } </div>
        <ImagesArea
          title={thisLaunch[0].mission_name}
          images={thisLaunch[0].links.flickr_images}
        />
      </Section>
    );
  }
}

LaunchView.propTypes = {
  launchCollection: PropTypes.shape({
    launches: PropTypes.arrayOf(PropTypes.shape({
      flight_number: PropTypes.string.isRequired,
      mission_name: PropTypes.string,
      launch_year: PropTypes.string,
      details: PropTypes.string,
      rocket: PropTypes.shape({
        rocket_id: PropTypes.string,
        rocket_name: PropTypes.string,
      }),
      links: PropTypes.shape({
        flickr_images: PropTypes.arrayOf(PropTypes.string),
      }),
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getLaunches: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      flight_number: PropTypes.string,
    }),
  }).isRequired,
}

const mapStateToProps = state => ({
  launchCollection: state.launchCollection,
});

const mapDispatchToProps = {
  getLaunches: getLaunchesThunk,
};

const ConnectedLaunchView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(LaunchView);

export default ConnectedLaunchView;
