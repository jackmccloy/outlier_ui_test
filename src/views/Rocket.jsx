import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getRocketsThunk } from "../actions/rockets";
import ImagesArea from '../components/ImagesArea';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const RocketH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class RocketView extends Component {
  componentDidMount() {
    const { getRockets } = this.props;
    getRockets();
  }

  render() {
    const {
      match,
      rocketCollection,
    } = this.props;

    if (!rocketCollection || rocketCollection.fetching) {
      return (
        <Section>
          <RocketH1> SpaceX </RocketH1>
          <div> LOADING... </div>
        </Section>
      );
    }

    if (rocketCollection.error) {
      return (
        <Section>
          <RocketH1> SpaceX </RocketH1>
          <div> { rocketCollection.error } </div>
        </Section>
      );
    }

    const rocketId = match.params.rocket_id;
    const thisRocket = rocketCollection.rockets.filter(rocket => rocket.rocket_id === rocketId);

    if (thisRocket.length !== 1) {
      return (
        <Section>
          <RocketH1> SpaceX </RocketH1>
          <div> NO DATA </div>
        </Section>
      );
    }

    return (
      <Section>
        <RocketH1> SpaceX { thisRocket[0].rocket_name } </RocketH1>
        <div> { thisRocket[0].description } </div>
        <ImagesArea
          title={thisRocket[0].rocket_name}
          images={thisRocket[0].flickr_images}
        />
      </Section>
    );
  }
}

RocketView.propTypes = {
  rocketCollection: PropTypes.shape({
    rockets: PropTypes.arrayOf(PropTypes.shape({
      rocket_id: PropTypes.string,
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getRockets: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      rocket_id: PropTypes.string,
    }),
  }).isRequired,
}

const mapStateToProps = state => ({
  rocketCollection: state.rocketCollection,
});

const mapDispatchToProps = {
  getRockets: getRocketsThunk,
};

const ConnectedRocketView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RocketView);

export default ConnectedRocketView;
