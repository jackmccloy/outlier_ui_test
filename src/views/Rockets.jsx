import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getRocketsThunk } from "../actions/rockets";
import Rocket from '../components/Rocket';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const RocketsH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class RocketsView extends Component {
  componentDidMount() {
    const { getRockets } = this.props;
    getRockets();
  }

  rocketsArea() {
    const { rocketCollection } = this.props;

    if (!rocketCollection || rocketCollection.fetching) {
      return (
        <div> LOADING... </div>
      );
    }

    if (rocketCollection.error) {
      return (
        <div> { rocketCollection.error } </div>
      );
    }

    if (!rocketCollection.rockets.length) {
      return (
        <div> NO DATA </div>
      );
    }

    const rockets = rocketCollection.rockets.map(rocket => (
      <Rocket
        key={rocket.rocket_id}
        rocket={rocket}
      />
    ));

    return (
      <ul>{rockets}</ul>
    );
  }

  render() {
    return (
      <Section>
        <RocketsH1> SpaceX rockets </RocketsH1>
        {this.rocketsArea()}
      </Section>
    );
  }
}

RocketsView.propTypes = {
  rocketCollection: PropTypes.shape({
    rockets: PropTypes.arrayOf(PropTypes.shape({
      rocket_id: PropTypes.string,
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getRockets: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  rocketCollection: state.rocketCollection,
});

const mapDispatchToProps = {
  getRockets: getRocketsThunk,
};

const ConnectedRocketsView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RocketsView);

export default ConnectedRocketsView;
