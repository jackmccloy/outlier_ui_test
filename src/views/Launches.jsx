import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getLaunchesThunk } from "../actions/Launches";
import Launch from '../components/Launch';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const LaunchesH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class LaunchesView extends Component {
  componentDidMount() {
    const { getLaunches } = this.props;
    getLaunches();
  }

  launchesArea() {
    const { launchCollection } = this.props;

    if (!launchCollection || launchCollection.fetching) {
      return <div> LOADING... </div>;
    }

    if (launchCollection.error) {
      return <div> { launchCollection.error } </div>
    }

    if (!launchCollection.launches.length) {
      return <div> NO DATA </div>;
    }

    const launches = launchCollection.launches.map(launch => (
      <Launch
        key={launch.flight_number}
        launch={launch}
      />
    ));

    return <ul>{launches}</ul>;
  }

  render() {
    return (
      <Section>
        <LaunchesH1> SpaceX launches </LaunchesH1>
        {this.launchesArea()}
      </Section>
    );
  }
}

LaunchesView.propTypes = {
  launchCollection: PropTypes.shape({
    launches: PropTypes.arrayOf(PropTypes.shape({
      flight_number: PropTypes.number,
      mission_name: PropTypes.string,
      rocket: PropTypes.shape({
        rocket_id: PropTypes.string,
        rocket_name: PropTypes.string,
      }),
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getLaunches: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  launchCollection: state.launchCollection,
});

const mapDispatchToProps = {
  getLaunches: getLaunchesThunk,
};

const ConnectedLaunchesView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(LaunchesView);

export default ConnectedLaunchesView;
