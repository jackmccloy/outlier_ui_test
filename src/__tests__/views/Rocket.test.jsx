import React from 'react';

import { mount } from 'enzyme';

import renderer from 'react-test-renderer';

import {
  MemoryRouter,
} from 'react-router';

import { RocketView } from '../../views/Rocket';

import { rocketApiResponse } from '../../__mocks__/apiResponses';

describe('<RocketView />', () => {
  let getRockets;

  beforeEach(() => {
    getRockets = jest.fn();

    mount(
      <RocketView
        getRockets={getRockets}
        rocketCollection={{
          rockets: [],
          fetching: false,
          error: '',
        }}
        match={{
          params: {
            rocket_id: 'falcon1',
          }
        }}
      />,
    );
  });

  it('calls getRockets on mount', () => {
    expect(getRockets.mock.calls.length).toBe(1);
  });

  it('renders correctly while fetching', () => {
    const tree = renderer
      .create(
        <RocketView
          getRockets={jest.fn()}
          rocketCollection={{
            rockets: [],
            fetching: true,
            error: '',
          }}
          match={{
            params: {
              rocket_id: 'falcon1',
            }
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly after fetch error', () => {
    const tree = renderer
      .create(
        <RocketView
          getRockets={jest.fn()}
          rocketCollection={{
            rockets: [],
            fetching: false,
            error: 'An error occured',
          }}
          match={{
            params: {
              rocket_id: 'falcon1',
            }
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with data', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <RocketView
            getRockets={jest.fn()}
            rocketCollection={{
              rockets: rocketApiResponse,
              fetching: false,
              error: '',
            }}
            match={{
              params: {
                rocket_id: 'falcon1',
              }
            }}
          />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
