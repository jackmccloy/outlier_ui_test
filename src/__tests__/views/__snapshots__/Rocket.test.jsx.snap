// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RocketView /> renders correctly after fetch error 1`] = `
<section
  className="Rocket__Section-sc-16sbiwh-0 eJKBUL"
>
  <h1
    className="Rocket__RocketH1-sc-16sbiwh-1 kWMeDo"
  >
     SpaceX 
  </h1>
  <div>
     
    An error occured
     
  </div>
</section>
`;

exports[`<RocketView /> renders correctly while fetching 1`] = `
<section
  className="Rocket__Section-sc-16sbiwh-0 eJKBUL"
>
  <h1
    className="Rocket__RocketH1-sc-16sbiwh-1 kWMeDo"
  >
     SpaceX 
  </h1>
  <div>
     LOADING... 
  </div>
</section>
`;

exports[`<RocketView /> renders correctly with data 1`] = `
<section
  className="Rocket__Section-sc-16sbiwh-0 eJKBUL"
>
  <h1
    className="Rocket__RocketH1-sc-16sbiwh-1 kWMeDo"
  >
     SpaceX 
    Falcon 1
     
  </h1>
  <div>
     
    The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
     
  </div>
  <div
    className="ImagesArea__ImageGridContainer-sc-172rgwy-0 jaRxtm"
  >
    <div
      className="ImagesArea__ImageContainer-sc-172rgwy-1 eWNNyX"
    >
      <a
        href="https://www.spacex.com/sites/spacex/files/styles/media_gallery_large/public/2009_-_01_liftoff_south_full_wide_ro8a1280_edit.jpg?itok=8loiSGt1"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="Falcon 1 1 of 2"
          src="https://www.spacex.com/sites/spacex/files/styles/media_gallery_large/public/2009_-_01_liftoff_south_full_wide_ro8a1280_edit.jpg?itok=8loiSGt1"
        />
      </a>
    </div>
    <div
      className="ImagesArea__ImageContainer-sc-172rgwy-1 eWNNyX"
    >
      <a
        href="https://www.spacex.com/sites/spacex/files/styles/media_gallery_large/public/2009_-_02_default_liftoff_west_full_wide_nn6p2062_xl.jpg?itok=p776nHsM"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="Falcon 1 2 of 2"
          src="https://www.spacex.com/sites/spacex/files/styles/media_gallery_large/public/2009_-_02_default_liftoff_west_full_wide_nn6p2062_xl.jpg?itok=p776nHsM"
        />
      </a>
    </div>
  </div>
</section>
`;
